docs/architecture-decisions/2023-01-12_smart-contract-deployment.md

```diff
@@ -0,0 +1,375 @@
+
+These different risk profiles have an impact on the functionality that is exposed within a smart contract. Two key examples of this are:
+
+- **Upgradeable smart contracts** - Whether or not a smart contract can be updated inline (and keep existing state and app ID / address characteristics) or they are immutable
+- **Deletable smart contracts** - Whether or not a smart contract can be deleted or is permanent
+
+Immutability and permanence are useful architectural properties in certain circumstances (similarly mutability and impermanence in others). For example:
+
+- If you have a smart contract that locks billions of dollars of value, then allowing that smart contract to be upgradeable allows for the manager of the smart contract to make a change that lets them steal the value.
+- If you have a smart contract that provides an Oracle service (say for betting odds), then allowing that smart contract to be deletable could break many other applications that are hard-coded to call that Oracle contract's app ID.
+- If you have a smart contract that runs the mechanics of a game engine the players of said game may have a reasonable expectation that the game engine is evolved and enhanced over time, but they don't want to lose their state (so upgrading the smart contract is useful).
+- If you have a smart contract that handles a one-off trade of value across a cross-chain bridge then it makes sense to delete it (only once trade has been concluded) to remove noise, potential confusion and operational overhead for the operators of said bridge.
+
+All 4 scenarios above provide different situations where immutability and permanence are desired or undesired. If you choose incorrectly for your circumstance, particularly for high-risk scenarios, the consequences could be major.
+
+If a contract is immutable, it does limit the ability to evolve the functionality over time based on user feedback and also discourages use of best practice software delivery techniques that encourage evolution and rapid deployment like Continuous Delivery.
+
+Another consideration in favour of immutability is smart contract auditing. If a smart contract is audited for security, but then the smart contract is upgradeable or is changed between when the audit occurred and when the contract was deployed to MainNet then the smart contract audit is somewhat invalidated and certainly any hashes that are provided of the smart contract code in an audit report will no longer match.
+
+There are techniques that can be used to allow for immutable smart contracts, but let them be evolved somewhat:
+
+- You could release a new version of a smart contract and include an ability for that smart contract to communicate to/from the existing smart contract to migrate state.
+- You could ensure that clients can dynamically find the latest smart contract so the application ID / address doesn't need to be hardcoded and the smart contract remains addressable.
```

Comment:
In case of oracles, another point of consideration is that different applications may have different requirements if the oracle disappear (which can happen for example if the data provider lost their keys or stop operating the oracle). Some more-security/decentralized-focus applications may decide to stop working. Some may decide to change oracle.

So I think it may make sense to systematically have such oracle be non-upgradable.
But then put a proxy smart contract in front that redirects to the right oracle for the applications that would like some form of updatability.

Reply:
Makes sense.